File: app/src/androidTest/java/info/firozansari/espressoexamples/ViewPagerActivityTests.java
package info.firozansari.espressoexamples;

import androidx.test.espresso.matcher.ViewMatchers;
import androidx.test.rule.ActivityTestRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;


import info.firozansari.espressoexamples.activities.ViewPagerActivity;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.swipeLeft;
import static androidx.test.espresso.action.ViewActions.swipeRight;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;

@RunWith (AndroidJUnit4.class)
public class ViewPagerActivityTests {

	/** Launches {@link ViewPagerActivity} for every test */
	@Rule
	public ActivityTestRule<ViewPagerActivity> activityRule = new ActivityTestRule<>(ViewPagerActivity.class);

	/**
     * Swipe between the first and second pages in the ViewPager.
     */
    @Test
    public void testSwipeBetweenFirstAndSecondPage() {
        onView(withText("Test One")).check(matches(isDisplayed()));
        onView(ViewMatchers.withId(R.id.pager)).perform(swipeLeft());
        onView(withText("Test Two")).check(matches(isDisplayed()));
    }

    /**
     * Swipe between the first and second pages and back to the first in the ViewPager.
     */
    @Test
    public void testSwipeBetweenFirstSecondAndBackToFirstPage() {
        onView(allOf(withId(R.id.example_text), withText("Test One"))).check(matches(isDisplayed()));
        onView(withId(R.id.pager)).perform(swipeLeft());
        onView(withText("Test Two")).check(matches(isDisplayed()));
        onView(withId(R.id.pager)).perform(swipeRight());
        onView(withText("Test One")).check(matches(isDisplayed()));
    }

    /**
     * Swipe to the end of a Fragment
     */
    @Test
    public void testSwipeToTheEnd() {
        onView(withText("Test One")).check(matches(isDisplayed()));
        onView(withId(R.id.pager)).perform(swipeLeft()).perform(swipeLeft()).perform(swipeLeft());
        onView(withText("Test Four")).check(matches(isDisplayed()));
    }

    /**
     * Swipe to the end of a Fragment and attempt to swipe further
     */
    @Test
    public void testSwipeBeyondTheEnd() {
        onView(withText("Test One")).check(matches(isDisplayed()));
        onView(withId(R.id.pager)).perform(swipeLeft()).perform(swipeLeft()).perform(swipeLeft()).perform(swipeLeft());
        onView(withText("Test Four")).check(matches(isDisplayed()));
    }
}

File: app/src/androidTest/java/info/firozansari/espressoexamples/ViewPagerActivityTests.kt
package info.firozansari.espressoexamples

import androidx.test.espresso.Espresso.*
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.internal.runner.junit4.statement.UiThreadStatement.runOnUiThread
import info.firozansari.espressoexamples.activities.MainActivity
import info.firozansari.espressoexamples.activities.ViewPagerActivity
import info.firozansari.espressoexamples.fragments.ViewPagerFragment
import org.hamcrest.Matchers
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ViewPagerActivityTests {

    @Rule
    @JvmField
    var activityRule = ActivityScenarioRule(MainActivity::class.java)

    @Before
    fun setup() {
        activityRule.scenario.onActivity{ activity ->
            runOnUiThread {
                activity.displayFragment(ViewPagerFragment())
            }
        }
    }

    /**
     * Swipe between the first and second pages in the ViewPager.
     */
    @Test
    fun testSwipeBetweenFirstAndSecondPage() {
        onView(ViewMatchers.withText("Test One")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withId(R.id.pager)).perform(ViewActions.swipeLeft())
        onView(ViewMatchers.withText("Test Two")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    /**
     * Swipe between the first and second pages and back to the first in the ViewPager.
     */
    @Test
    fun testSwipeBetweenFirstSecondAndBackToFirstPage() {
        onView(Matchers.allOf(ViewMatchers.withId(R.id.example_text), ViewMatchers.withText("Test One"))).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withId(R.id.pager)).perform(ViewActions.swipeLeft())
        onView(ViewMatchers.withText("Test Two")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withId(R.id.pager)).perform(ViewActions.swipeRight())
        onView(ViewMatchers.withText("Test One")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    /**
     * Swipe to the end of a Fragment
     */
    @Test
    fun testSwipeToTheEnd() {
        onView(ViewMatchers.withText("Test One")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withId(R.id.pager)).perform(ViewActions.swipeLeft()).perform(ViewActions.swipeLeft()).perform(ViewActions.swipeLeft())
        onView(ViewMatchers.withText("Test Four")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    /**
     * Swipe to the end of a Fragment and attempt to swipe further
     */
    @Test
    fun testSwipeBeyondTheEnd() {
        onView(ViewMatchers.withText("Test One")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        onView(ViewMatchers.withId(R.id.pager)).perform(ViewActions.swipeLeft()).perform(ViewActions.swipeLeft()).perform(ViewActions.swipeLeft()).perform(ViewActions.swipeLeft())
        onView(ViewMatchers.withText("Test Four")).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }
}